File: java/com/example/spottio/AddPostFragment.java
package com.example.spottio;

import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.AutoCompleteTextView;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.viewpager2.widget.ViewPager2;

import java.util.ArrayList;
import java.util.List;

public class AddPostFragment extends Fragment {

    private Uri selectedMediaUri = null;
    private boolean isVideoSelected = false;
    private ImageView ivPreview;
    private AutoCompleteTextView autoCompleteCategory;

    private List<String> categoryList = new ArrayList<>();
    private ArrayAdapter<String> adapter;
    private AddPostHandler addPostHandler;

    private final ActivityResultLauncher<String> galleryLauncher = registerForActivityResult(
            new ActivityResultContracts.GetContent(),
            uri -> {
                if (uri != null) {
                    selectedMediaUri = uri;
                    String type = requireContext().getContentResolver().getType(uri);
                    isVideoSelected = type != null && type.startsWith("video");
                    ivPreview.setImageURI(uri);
                    ivPreview.setVisibility(View.VISIBLE);
                }
            }
    );

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_add_post, container, false);

        addPostHandler = new AddPostHandler();

        EditText etText = view.findViewById(R.id.etPostText);
        Button btnMedia = view.findViewById(R.id.btnAttachMedia);
        Button btnPublish = view.findViewById(R.id.btnPublish);
        ivPreview = view.findViewById(R.id.ivMediaPreview);
        autoCompleteCategory = view.findViewById(R.id.autoCompleteCategory);

        adapter = new ArrayAdapter<>(getContext(), android.R.layout.simple_spinner_item, categoryList);
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        autoCompleteCategory.setAdapter(adapter);

        fetchCategories();

        btnMedia.setOnClickListener(v -> galleryLauncher.launch("*/*"));

        btnPublish.setOnClickListener(v -> handlePublish(etText, autoCompleteCategory.getText().toString()));

        return view;
    }

    private void fetchCategories() {
        addPostHandler.loadCategories(new AddPostHandler.OnCategoriesLoadedListener() {
            @Override
            public void onSuccess(List<String> categories) {
                categoryList.clear();
                categoryList.addAll(categories);
                adapter.notifyDataSetChanged();
            }

            @Override
            public void onFailure(Exception e) {
                Log.e("Firestore", "Errore nel caricamento categorie", e);
            }
        });
    }

    private void handlePublish(EditText etText, String category) {
        String text = etText.getText().toString().trim();

        // Verifica che l'utente non stia tentando di pubblicare un post completamente vuoto
        if (text.isEmpty() && selectedMediaUri == null) {
            Toast.makeText(getContext(), "Scrivi qualcosa o allega un file!", Toast.LENGTH_SHORT).show();
            return;
        }

        // Recupero UID centralizzato e sicuro tramite AuthManager
        String currentUid = AuthManager.getCurrentUserUid(requireContext());

        // Blocco di sicurezza nel caso in cui la sessione sia scaduta o non valida
        if (currentUid.isEmpty()) {
            Toast.makeText(getContext(), "Errore: sessione non valida.", Toast.LENGTH_SHORT).show();
            return;
        }

        String mediaUriString = (selectedMediaUri != null) ? selectedMediaUri.toString() : "";

        // Validazione e assegnazione di un valore di default alla categoria
        String finalCategory = category.trim();
        if (finalCategory.isEmpty()) {
            finalCategory = "Generale"; // Evita l'invio di stringhe vuote a Firestore
        }

        // Creazione dell'oggetto Post con i parametri verificati e puliti
        Post newPost = new Post(currentUid, text, mediaUriString, isVideoSelected, finalCategory);

        addPostHandler.publishPost(newPost, new AddPostHandler.OnPostPublishedListener() {
            @Override
            public void onSuccess() {
                Toast.makeText(getContext(), "Post pubblicato!", Toast.LENGTH_SHORT).show();
                etText.setText("");
                ivPreview.setVisibility(View.GONE);
                selectedMediaUri = null;

                // Ritorna automaticamente alla tab principale del feed (HomeFragment)
                ViewPager2 vp = requireActivity().findViewById(R.id.viewPager);
                if (vp != null) vp.setCurrentItem(0, true);
            }

            @Override
            public void onFailure(Exception e) {
                Toast.makeText(getContext(), "Errore pubblicazione", Toast.LENGTH_SHORT).show();
                Log.e("Firestore", "Err: " + e.getMessage());
            }
        });
    }
}

File: app/src/main/java/com/example/spottio/AddPostFragment.kt
package com.example.spottio

import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.AutoCompleteTextView
import android.widget.Button
import android.widget.EditText
import android.widget.ImageView
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.fragment.app.Fragment
import androidx.viewpager2.widget.ViewPager2
import com.google.firebase.firestore.FirebaseFirestore

class AddPostFragment : Fragment() {

    private var selectedMediaUri: Uri? = null
    private var isVideoSelected = false

    private lateinit var ivPreview: ImageView
    private lateinit var autoCompleteCategory: AutoCompleteTextView
    private lateinit var etText: EditText

    private val categoryList = mutableListOf<String>()
    private lateinit var adapter: ArrayAdapter<String>

    // Inizializza Firestore direttamente qui (AddPostHandler non serve più)
    private val db = FirebaseFirestore.getInstance()

    private val galleryLauncher = registerForActivityResult(ActivityResultContracts.GetContent()) { uri: Uri? ->
        if (uri != null) {
            selectedMediaUri = uri
            // Uso context?. per evitare crash se il fragment viene chiuso mentre si sceglie il file
            val type = context?.contentResolver?.getType(uri)
            isVideoSelected = type != null && type.startsWith("video")
            ivPreview.setImageURI(uri)
            ivPreview.visibility = View.VISIBLE
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.fragment_add_post, container, false)

        etText = view.findViewById(R.id.etPostText)
        val btnMedia = view.findViewById<Button>(R.id.btnAttachMedia)
        val btnPublish = view.findViewById<Button>(R.id.btnPublish)
        ivPreview = view.findViewById(R.id.ivMediaPreview)
        autoCompleteCategory = view.findViewById(R.id.autoCompleteCategory)

        // Setup Adapter in puro stile Kotlin
        adapter = ArrayAdapter(requireContext(), android.R.layout.simple_spinner_item, categoryList).apply {
            setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        }
        autoCompleteCategory.setAdapter(adapter)

        fetchCategories()

        btnMedia.setOnClickListener { galleryLauncher.launch("*/*") }

        // Non c'è bisogno di passare etText perché è già una variabile di classe
        btnPublish.setOnClickListener { handlePublish(autoCompleteCategory.text.toString()) }

        return view
    }

    private fun fetchCategories() {
        db.collection("categories")
            .orderBy("name")
            .get()
            .addOnCompleteListener { task ->
                if (task.isSuccessful && task.result != null) {
                    categoryList.clear()
                    for (document in task.result!!) {
                        document.getString("name")?.let { categoryList.add(it) }
                    }
                    // Aggiorniamo l'interfaccia solo se il fragment è ancora aperto
                    if (isAdded) {
                        adapter.notifyDataSetChanged()
                    }
                } else {
                    Log.e("Firestore", "Errore nel caricamento categorie", task.exception)
                }
            }
    }

    private fun handlePublish(category: String) {
        val text = etText.text.toString().trim()

        if (text.isEmpty() && selectedMediaUri == null) {
            Toast.makeText(context, "Scrivi qualcosa o allega un file!", Toast.LENGTH_SHORT).show()
            return
        }

        // Recupero sicuro del Context per evitare crash con RequireContext()
        val safeContext = context ?: return
        val currentUid = AuthManager.getCurrentUserUid(safeContext)

        if (currentUid.isEmpty()) {
            Toast.makeText(safeContext, "Errore: sessione non valida.", Toast.LENGTH_SHORT).show()
            return
        }

        val mediaUriString = selectedMediaUri?.toString() ?: ""

        var finalCategory = category.trim()
        if (finalCategory.isEmpty()) {
            finalCategory = "Generale"
        }

        val newPost = Post(currentUid, text, mediaUriString, isVideoSelected, finalCategory)

        db.collection("posts")
            .add(newPost)
            .addOnSuccessListener {
                Toast.makeText(context, "Post pubblicato!", Toast.LENGTH_SHORT).show()

                // Svuotiamo il testo nel modo corretto in Kotlin per evitare il warning del setText
                etText.text.clear()
                ivPreview.visibility = View.GONE
                selectedMediaUri = null

                // Cambio pagina sicuro (se l'activity esiste ancora)
                activity?.findViewById<ViewPager2>(R.id.viewPager)?.setCurrentItem(0, true)
            }
            .addOnFailureListener { e ->
                Toast.makeText(context, "Errore pubblicazione", Toast.LENGTH_SHORT).show()
                Log.e("Firestore", "Err: ${e.message}")
            }
    }
}